File: backend/src/scripts/update-usdt-inr-rate.ts
/**
 * Cron Script: Update USDT/INR Conversion Rate
 *
 * This script fetches the current USDT/INR conversion rate from CoinDCX
 * and stores it in the SystemConfig table.
 *
 * Run frequency: Every 4 months (or manually when needed)
 *
 * Usage:
 *   npx ts-node src/scripts/update-usdt-inr-rate.ts
 *   # or after building:
 *   node dist/scripts/update-usdt-inr-rate.js
 */

import crypto from 'crypto';
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

const COINDCX_API_URL = 'https://api.coindcx.com/api/v1/derivatives/futures/data/conversions';

interface ConversionData {
  symbol: string;
  margin_currency_short_name: string;
  target_currency_short_name: string;
  conversion_price: number;
  last_updated_at: number;
}

async function fetchConversionRate(apiKey: string, apiSecret: string): Promise<number | null> {
  const timestamp = Date.now();
  const body = { timestamp };

  const payload = Buffer.from(JSON.stringify(body)).toString();
  const signature = crypto.createHmac('sha256', apiSecret).update(payload).digest('hex');

  try {
    const response = await fetch(COINDCX_API_URL, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-AUTH-APIKEY': apiKey,
        'X-AUTH-SIGNATURE': signature,
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      console.error(`API request failed: ${response.status} ${response.statusText}`);
      return null;
    }

    const data = (await response.json()) as ConversionData[];
    const usdtInr = data.find(item => item.symbol === 'USDTINR');

    if (usdtInr) {
      console.log(`Fetched USDT/INR rate: ${usdtInr.conversion_price}`);
      console.log(`Last updated at: ${new Date(usdtInr.last_updated_at).toISOString()}`);
      return usdtInr.conversion_price;
    }

    console.error('USDTINR not found in response');
    return null;
  } catch (error) {
    console.error('Failed to fetch conversion rate:', error);
    return null;
  }
}

async function main() {
  console.log('='.repeat(60));
  console.log('USDT/INR Rate Update Script');
  console.log('='.repeat(60));
  console.log(`Started at: ${new Date().toISOString()}`);
  console.log('');

  try {
    // Pick a random active broker credential
    const credentials = await prisma.brokerCredential.findMany({
      where: { isActive: true },
      select: { apiKey: true, apiSecret: true },
    });

    if (credentials.length === 0) {
      console.error('No active broker credentials found in database');
      process.exit(1);
    }

    // Pick a random one
    const randomIndex = Math.floor(Math.random() * credentials.length);
    const { apiKey, apiSecret } = credentials[randomIndex];
    console.log(`Using credential ${randomIndex + 1} of ${credentials.length} (randomly selected)`);

    // Fetch the conversion rate
    const rate = await fetchConversionRate(apiKey, apiSecret);

    if (rate === null) {
      console.error('Failed to fetch conversion rate');
      process.exit(1);
    }

    // Store in database
    await prisma.systemConfig.upsert({
      where: { key: 'USDT_INR_RATE' },
      update: { value: rate.toString() },
      create: { key: 'USDT_INR_RATE', value: rate.toString() },
    });

    console.log('');
    console.log(`✅ Successfully updated USDT_INR_RATE to ${rate}`);
    console.log(`Stored in system_config table`);
    console.log('');
    console.log('Next update recommended in 4 months');
    console.log('='.repeat(60));

  } catch (error) {
    console.error('Script failed:', error);
    process.exit(1);
  } finally {
    await prisma.$disconnect();
  }
}

main();

File: backend/src/scripts/update-usdt-inr-rate.py
#!/usr/bin/env python3
"""
Cron Script: Update USDT/INR Conversion Rate

This script fetches the current USDT/INR conversion rate from CoinDCX
and stores it in the SystemConfig table.

Run frequency: Every 4 months (or manually when needed)

Usage:
    python3 src/scripts/update-usdt-inr-rate.py

Requirements:
    pip install requests psycopg2-binary
"""

import hmac
import hashlib
import json
import time
import os
import sys
import random
import requests
import psycopg2
from urllib.parse import urlparse

COINDCX_API_URL = "https://api.coindcx.com/api/v1/derivatives/futures/data/conversions"


def get_database_url():
    """Get database URL from environment or .env file"""
    db_url = os.environ.get("DATABASE_URL")
    if db_url:
        return db_url

    # Try to read from .env file
    env_path = os.path.join(os.path.dirname(__file__), "../../.env")
    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                if line.startswith("DATABASE_URL="):
                    return line.strip().split("=", 1)[1].strip('"')

    raise Exception("DATABASE_URL not found")


def fetch_conversion_rate(api_key: str, api_secret: str) -> float | None:
    """Fetch USDT/INR conversion rate from CoinDCX API"""
    timestamp = int(round(time.time() * 1000))
    body = {"timestamp": timestamp}
    json_body = json.dumps(body, separators=(",", ":"))

    signature = hmac.new(
        api_secret.encode("utf-8"),
        json_body.encode("utf-8"),
        hashlib.sha256
    ).hexdigest()

    headers = {
        "Content-Type": "application/json",
        "X-AUTH-APIKEY": api_key,
        "X-AUTH-SIGNATURE": signature
    }

    try:
        response = requests.get(COINDCX_API_URL, data=json_body, headers=headers)
        response.raise_for_status()

        data = response.json()
        for item in data:
            if item.get("symbol") == "USDTINR":
                rate = item.get("conversion_price")
                last_updated = item.get("last_updated_at")
                print(f"Fetched USDT/INR rate: {rate}")
                print(f"Last updated at: {time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(last_updated / 1000))}")
                return float(rate)

        print("USDTINR not found in response")
        return None

    except Exception as e:
        print(f"Failed to fetch conversion rate: {e}")
        return None


def main():
    print("=" * 60)
    print("USDT/INR Rate Update Script")
    print("=" * 60)
    print(f"Started at: {time.strftime('%Y-%m-%d %H:%M:%S')}")
    print()

    try:
        # Get database connection
        db_url = get_database_url()
        parsed = urlparse(db_url)

        conn = psycopg2.connect(
            host=parsed.hostname,
            port=parsed.port or 5432,
            user=parsed.username,
            password=parsed.password,
            database=parsed.path[1:]  # Remove leading /
        )
        cursor = conn.cursor()

        # Pick a random active broker credential
        cursor.execute('''
            SELECT "apiKey", "apiSecret"
            FROM broker_credentials
            WHERE "isActive" = true
            ORDER BY RANDOM()
            LIMIT 1
        ''')
        row = cursor.fetchone()

        if not row:
            print("No active broker credentials found in database")
            sys.exit(1)

        api_key, api_secret = row
        print(f"Using a randomly selected credential")

        # Fetch the conversion rate
        rate = fetch_conversion_rate(api_key, api_secret)

        if rate is None:
            print("Failed to fetch conversion rate")
            sys.exit(1)

        # Store in database
        cursor.execute('''
            INSERT INTO system_config (key, value, "updatedAt")
            VALUES ('USDT_INR_RATE', %s, NOW())
            ON CONFLICT (key) DO UPDATE
            SET value = %s, "updatedAt" = NOW()
        ''', (str(rate), str(rate)))

        conn.commit()

        print()
        print(f"✅ Successfully updated USDT_INR_RATE to {rate}")
        print("Stored in system_config table")
        print()
        print("Next update recommended in 4 months")
        print("=" * 60)

        cursor.close()
        conn.close()

    except Exception as e:
        print(f"Script failed: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
